phpBB/includes/acp/acp_update.php

```diff
@@ -49,40 +89,235 @@ function main($id, $mode)
 
-			'UPDATE_INSTRUCTIONS'		=> $user->lang('UPDATE_INSTRUCTIONS', $update_link),
+			'UPDATE_INSTRUCTIONS'		=> $this->language->lang('UPDATE_INSTRUCTIONS', "{$this->root_path}install/app.{$this->php_ext}"),
 			'S_VERSION_UPGRADEABLE'		=> !empty($upgrades_available),
-			'UPGRADE_INSTRUCTIONS'		=> !empty($upgrades_available) ? $user->lang('UPGRADE_INSTRUCTIONS', $upgrades_available['current'], $upgrades_available['announcement']) : false,
-		];
-
-		$template->assign_vars($template_ary);
+			'UPGRADE_INSTRUCTIONS'		=> !empty($upgrades_available) ? $this->language->lang('UPGRADE_INSTRUCTIONS', $upgrades_available['current'], $upgrades_available['announcement']) : false,
+		]);
 
 		// Incomplete update?
-		if (phpbb_version_compare($config['version'], PHPBB_VERSION, '<'))
+		if ($version_helper->compare($config['version'], PHPBB_VERSION, '<'))
 		{
-			$database_update_link = $phpbb_root_path . 'install/app.php/update';
-
-			$template->assign_vars(array(
+			$template->assign_vars([
 				'S_UPDATE_INCOMPLETE'		=> true,
 				'FILES_VERSION'				=> PHPBB_VERSION,
-				'INCOMPLETE_INSTRUCTIONS'	=> $user->lang('UPDATE_INCOMPLETE_EXPLAIN', $database_update_link),
-			));
+				'INCOMPLETE_INSTRUCTIONS'	=> $language->lang('UPDATE_INCOMPLETE_EXPLAIN', "{$this->root_path}install/app.{$this->php_ext}/update"),
+			]);
 		}
 	}
+
+	/**
+	 * Perform an instant update.
+	 *
+	 * @param array		$update		Update data
+	 * @return void
+	 */
+	protected function instant_update(array $update): void
+	{
+		if ($this->check_english_language() === false)
+		{
+			$this->instant_update_error($this->language->lang('INSTANT_UPDATE_ERROR_ENGLISH_REQUIRED'));
+		}
+
+		/** @var \phpbb\filesystem\filesystem $filesystem */
+		$filesystem = $this->container->get('filesystem');
+
+		// Ensure that the store directory is writable
+		if ($filesystem->is_writable($this->root_path . 'store') === false)
+		{
+			$this->instant_update_error(
+				$this->language->lang('DIRECTORY_NOT_WRITABLE') . '<br>' .
+				$this->language->lang('MAKE_FOLDER_WRITABLE', '/store')
+			);
+		}
+
+		$temp_dir = "{$this->root_path}store/phpbb-update-{$update['current']}";
+		$temp_zip = "{$temp_dir}.zip";
+
+		// Make sure the temporary directories and files do not already exist
+		$filesystem->remove([$temp_dir, $temp_zip]);
+
+		try
+		{
+			// Download the new version's package
+			(new \GuzzleHttp\Client)->request('GET', $update['download'], ['sink' => $temp_zip]);
+		}
+		catch (\GuzzleHttp\Exception\GuzzleException $e)
+		{
+			$filesystem->remove($temp_zip);
+
+			$this->instant_update_error($this->language->lang('INSTANT_UPDATE_ERROR_DOWNLOAD', $e->getMessage()));
+		}
+
+		// Verify the sha256 checksum of the download
+		if ($update['checksum'] !== hash_file('sha256', $filesystem->realpath($temp_zip)))
+		{
+			$filesystem->remove($temp_zip);
+
+			$this->instant_update_error($this->language->lang('INSTANT_UPDATE_ERROR_CHECKSUM_MISMATCH'));
+		}
+
+		$zip = new \ZipArchive;
```

Comment:
Add ext-zip to composer.json or better, use the phpbb zip implementation in functions_compress.php